components/layout_2020/display_list/mod.rs

```diff
@@ -893,7 +902,91 @@ impl<'a> BuilderForBoxFragment<'a> {
         }
     }
 
+    fn build_collapsed_table_borders(&mut self, builder: &mut DisplayListBuilder) {
+        let table_info = match &self.fragment.detailed_layout_info {
+            Some(SpecificLayoutInfo::TableGridWithCollapsedBorders(table_info)) => table_info,
+            _ => return,
+        };
+        let mut common =
+            builder.common_properties(units::LayoutRect::default(), &self.fragment.style);
+        let radius = wr::BorderRadius::default();
+        let mut column_sum = Au::zero();
+        for (x, column_size) in table_info.track_sizes.x.iter().enumerate() {
+            let mut row_sum = Au::zero();
+            for (y, row_size) in table_info.track_sizes.y.iter().enumerate() {
+                let left_border = &table_info.collapsed_borders.x[x].list[y];
+                let right_border = &table_info.collapsed_borders.x[x + 1].list[y];
+                let top_border = &table_info.collapsed_borders.y[y].list[x];
+                let bottom_border = &table_info.collapsed_borders.y[y + 1].list[x];
+                let border_widths = PhysicalSides::new(
+                    if y == 0 { top_border.width } else { Au::zero() },
+                    right_border.width,
+                    bottom_border.width,
+                    if x == 0 {
+                        left_border.width
+                    } else {
+                        Au::zero()
+                    },
+                );
+                let details = wr::BorderDetails::Normal(wr::NormalBorder {
+                    left: self.build_border_side(left_border.style_color.clone()),
+                    right: self.build_border_side(right_border.style_color.clone()),
+                    top: self.build_border_side(top_border.style_color.clone()),
+                    bottom: self.build_border_side(bottom_border.style_color.clone()),
+                    radius,
+                    do_aa: true,
+                });
+
+                let mut origin = PhysicalPoint::new(column_sum, row_sum);
+                let mut size = PhysicalSize::new(*column_size, *row_size);
+                if x == 0 {
+                    origin.x -= table_info.wrapper_border.left;
+                    size.width += table_info.wrapper_border.left;
+                } else {
+                    origin.x += left_border.width / 2;
+                    size.width -= left_border.width / 2;
+                }
+                if y == 0 {
+                    origin.y -= table_info.wrapper_border.top;
+                    size.height += table_info.wrapper_border.top;
+                } else {
+                    origin.y += top_border.width / 2;
+                    size.height -= top_border.width / 2;
+                }
```

Comment:
If you are already mutating `origin` and `size` here maybe it makes sense to mutate `border_widths` in the same way for the `y != 0` and x != 0` cases?